drivers/ieee802154/ieee802154_kw41z.c

```diff
@@ -425,130 +660,217 @@ static int kw41z_tx(struct device *dev, struct net_pkt *pkt,
+	if (KW41Z_AUTOACK_ENABLED && payload->fc.ar) {
+		tx_timeout = kw41z->tx_warmup_time + KW41Z_SHR_PHY_TIME +
+				 payload_len * KW41Z_PER_BYTE_TIME + 10 +
+				 KW41Z_ACK_WAIT_TIME;
+
+		SYS_LOG_DBG("AUTOACK_ENABLED: len: %d, timeout: %d, seq: %d",
+			payload_len, tx_timeout, payload->sequence);
+
+		kw41z_tmr3_set_timeout(tx_timeout);
 		ZLL->PHY_CTRL |= ZLL_PHY_CTRL_RXACKRQD_MASK;
 		kw41z_set_seq_state(KW41Z_STATE_TXRX);
-		kw41z_tmr2_set_timeout(tx_timeout);
-	} else {
+	} else
+#endif
+	{
+		SYS_LOG_DBG("AUTOACK disabled: len: %d, seq: %d",
+			payload_len, payload->sequence);
+
 		ZLL->PHY_CTRL &= ~ZLL_PHY_CTRL_RXACKRQD_MASK;
 		kw41z_set_seq_state(KW41Z_STATE_TX);
-		kw41z_tmr1_set_timeout(tx_timeout);
 	}
 
 	kw41z_enable_seq_irq();
 
 	k_sem_take(&kw41z->seq_sync, K_FOREVER);
 
+	SYS_LOG_DBG("seq_retval: %d", kw41z->seq_retval);
 	return kw41z->seq_retval;
 }
 
 static void kw41z_isr(int unused)
 {
 	u32_t irqsts = ZLL->IRQSTS;
 	u8_t state = kw41z_get_seq_state();
+	u32_t seq_state = ZLL->SEQ_STATE;
 	u8_t restart_rx = 1;
-	u8_t rx_len;
-
-	/* TMR1 IRQ - time-out */
-	if ((irqsts & ZLL_IRQSTS_TMR1IRQ_MASK) &&
-	    !(irqsts & ZLL_IRQSTS_TMR1MSK_MASK)) {
-		SYS_LOG_DBG("TMR1 timeout");
-		kw41z_tmr1_disable();
-		kw41z_disable_seq_irq();
-
-		if (state == KW41Z_STATE_CCA &&
-		    !(irqsts & ZLL_IRQSTS_CCA_MASK)) {
-			kw41z_set_seq_state(KW41Z_STATE_IDLE);
-			atomic_set(&kw41z_context_data.seq_retval, 0);
-			restart_rx = 0;
-		} else {
-			atomic_set(&kw41z_context_data.seq_retval, -EBUSY);
-		}
+	u32_t rx_len;
 
-		k_sem_give(&kw41z_context_data.seq_sync);
-	}
+	SYS_LOG_DBG("irqsts: 0x%08X, PHY_CTRL: 0x%08X, "
+			"SEQ_STATE: 0x%08X, state: %d",
+			irqsts, (unsigned int)ZLL->PHY_CTRL,
+			(unsigned int)seq_state, state);
+
+	/* Clear interrupts */
+	ZLL->IRQSTS = irqsts;
 
-	/* TMR2 IRQ - time-out */
-	if ((irqsts & ZLL_IRQSTS_TMR2IRQ_MASK) &&
-	    !(irqsts & ZLL_IRQSTS_TMR2MSK_MASK)) {
-		SYS_LOG_DBG("TMR2 timeout");
-		kw41z_tmr2_disable();
-		atomic_set(&kw41z_context_data.seq_retval, -EBUSY);
-		k_sem_give(&kw41z_context_data.seq_sync);
+	if ((!(ZLL->PHY_CTRL & ZLL_PHY_CTRL_RX_WMRK_MSK_MASK)) &&
+	    (irqsts & ZLL_IRQSTS_RXWTRMRKIRQ_MASK)) {
+		/*
+		 * There is a bug in the KW41Z where in noisy environments
+		 * the RX sequence can get lost. The watermark mask IRQ can
+		 * start TMR3 to complete the rest of the read or to assert
+		 * IRQ if the sequencer gets lost so we can reset things.
+		 * Note that a TX from the upper layers will also reset
+		 * things so the problem is contained a bit in normal
+		 * operation.
+		 */
+		rx_len = (irqsts & ZLL_IRQSTS_RX_FRAME_LENGTH_MASK)
+			>> ZLL_IRQSTS_RX_FRAME_LENGTH_SHIFT;
+
+		SYS_LOG_DBG("WMRK irq: seq_state: 0x%08x, rx_len: %d",
+			seq_state, rx_len);
+
+		KW_DBG_TRACE(KW41_DBG_TRACE_WTRM, irqsts,
+			(unsigned int)ZLL->PHY_CTRL, seq_state);
+
+		/* Convert to symbols and add IFS and ACK duration */
+		rx_len = rx_len * 2 + 12 + 22 + 2;
```

Comment:
do we have any defines for the 12, 22, 2?  If not maybe be clear in the comment above about which num is what.